SLING-10502 - lazy loading helpers

bdelacretaz opened the pull request:
This is an initial implementation with a `LazyLoadingField`and `LazyLoadingMap` that should help build GraphQL result sets with lazy loading values.

jsedding commented on src/main/java/org/apache/sling/graphql/helpers/layzloading/LazyLoadingMap.java:
```diff
@@ -0,0 +1,177 @@
+     */
+    private void computeAll() {
+        log.info("computeAll called, all remaining lazy values will be evaluated now");
+        suppliers.entrySet().forEach(e -> {
```
Does `suppliers.keySet().forEach(key -> this.get(key))` also work? Or does it cause `ConcurrentmodificationException` because `suppliers` has its elements removed during iteration? If that's the case, copying `suppliers.keySet()` before iteration should work.

On a related note: I think the current implementation needs to call `suppliers.clear()` to be correct.

raducotescu reviewed (changes requested):
I agree with all the observations that @jsedding made. There's no need to synchronize the `LazyLoadingMap` if it extends `HashMap`. Leave the synchronization to the caller - they can always wrap it in a `Collections#synchronizedMap`, or extend `ConcurrentHashMap`.

bdelacretaz commented:
Thank you for the detailed review, all good suggestions! Of course, HashMap not being synchronized it doesn't make sense to make the `LazyLoadingMap` thread safe, I have removed all synchronization in that class.

I think I have implemented all other suggestions, please cross-check.

jsedding reviewed (approved):
LGTM!

raducotescu commented on src/main/java/org/apache/sling/graphql/helpers/layzloading/LazyLoadingMap.java:
```diff
@@ -0,0 +1,176 @@
+    @Override
+    public T remove(Object key) {
+        super.remove(key);
+        suppliers.remove(key);
```
Would it be too expensive to obey the `Map` contract in this case?

bdelacretaz replied:
It's hard to say - I will make that optional, adding an `Options` object to the class constructor.

bdelacretaz replied:
Committed that, using class methods to set options instead.

raducotescu commented on src/main/java/org/apache/sling/graphql/helpers/layzloading/LazyLoadingMap.java:
```diff
@@ -0,0 +1,176 @@
+     */
+    private void computeAll() {
+        if(!suppliers.isEmpty()) {
+            log.info("computeAll called, all remaining lazy values will be evaluated now");
```
I think this should be logged on `debug` - I assume that with GraphQL in Sling catching on, we'll have a lot of `LazyLoadingMaps`.

bdelacretaz replied:
+1 I'll do that

jsedding commented on src/main/java/org/apache/sling/graphql/helpers/lazyloading/LazyLoadingMap.java:
```diff
@@ -0,0 +1,197 @@
+        }
+        super.remove(key);
+        suppliers.remove(key);
+        return null;
```
@bdelacretaz I think you meant to return `super.remove(key)` here if `computeValueOnRemove` is `true`. How about this?

```
Supplier<T> supplier = suppliers.remove(key);
if (supplier != null) {
    super.remove(key); // this should be unnecessary, otherwise I believe it would be a bug?!
    return computeValueOnRemove ? supplier.get() : null;
} else {
    return super.remove(key);    
}
```

bdelacretaz replied:
Indeed, _both the tests and the code were wrong_!

Good idea to use the Supplier in this way, I have committed a slightly different fix that sets the Stats and that always returns null unless `computeValueOnRemove` is set - I think that's less confusing. https://github.com/apache/sling-org-apache-sling-graphql-core/commit/4cce0790dea00fb5109606531817f17304f62450

jsedding replied:
👍 looks good to me
